text/0000-rc-newref-clone.md

```diff
@@ -0,0 +1,80 @@
+- Feature Name: refcount_clone_new_ref
+- Start Date: 2017-03-17
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+This RFC proposes the simple addition of a new_ref method to reference-counted pointer types (Rc, Arc and repsective weak pointer types) in the standard library synonymous to clone, with a more explicit terminology.
+
+# Motivation
+[motivation]: #motivation
+
+Clone is one of the most vaguely defined concept in rust. Clone implementations can be shallow or deep clones which, as the [standard libray documentation](https://doc.rust-lang.org/std/clone/trait.Clone.html) puts it "may or may not be expensive".
+
+It is therefore required, when reading code that calls clone to know the exact type of the calling object, because the cloning a ```Vec<T>``` and cloning an ```Arc<Vec<T>>``` have very different logic and performance implications, while their semantic is really mostly driven by the type of ```T```.
+
+This leads to recurrent misunderstandings when reading or reviewing code (See the example section at the end of this document).
+In many cases the code could be a lot easier to understand at a glance if reference counted pointers had a method with a more explicit name to create new reference.
+
+This issue is not specific to clone. The author of this RFC believes that it is generally best to avoid using a high level and vaguely defined concept as the only way to name an operation. ```Clone::clone``` is a prime example because it is at the same time an intentionally vague concept, can imply either a very cheap or a very costly operation, and can go from one to the other with simple changes to the code that don't necessarily affect the overal semantic of the progam. This RFC therefore concentrates on the case of Clone and reference counted types in the standard library.
+
+# Detailed design
+[design]: #detailed-design
+
+This RFC does not involve any compiler change (only minor additions to alloc/rc.rs and alloc/arc.rs).
+
+The following steps apply to ```Rc<T>```, ```Arc<T>```, ```rc::Weal<T>```, ```arc::Weak<T>```.
+
+A method ```fn new_ref(&self) -> Self``` is added to the pointer type, into which the code of the ```Clone::clone``` implementation is moved.
```

Comment:
Better define it as

```rust
fn new_ref(this: &Self) -> Self;
```

to let user call it as

```rust
let p1 = Rc::new(vec![0u8; 1000]);
let p2 = Rc::new_ref(&p1);   // <-- instead of p1.new_ref()
```

as explained in the doc of Rc:

> The inherent methods of `Rc` are all associated functions, which means that you have to call them as e.g. `Rc::get_mut(&value)` instead of `value.get_mut()`. This avoids conflicts with methods of the inner type `T`.